server/src/main/java/org/elasticsearch/repositories/IndexMetaDataGenerations.java

```diff
@@ -0,0 +1,174 @@
+        final Map<SnapshotId, Map<IndexId, String>> updatedIndexMetaLookup = new HashMap<>(this.lookup);
+        final Map<String, String> updatedIndexMetaIdentifiers = new HashMap<>(identifiers);
+        updatedIndexMetaIdentifiers.putAll(newIdentifiers);
+        updatedIndexMetaLookup.compute(snapshotId, (snId, lookup) -> {
+            if (lookup == null) {
+                if (newLookup.isEmpty()) {
+                    return null;
+                }
+                return Map.copyOf(newLookup);
+            } else {
+                final Map<IndexId, String> updated = new HashMap<>(lookup);
+                updated.putAll(newLookup);
+                return Map.copyOf(updated);
+            }
+        });
+        return new IndexMetaDataGenerations(updatedIndexMetaLookup, updatedIndexMetaIdentifiers);
+    }
+
+    /**
+     * Create a new instance with the given snapshot removed.
+     *
+     * @param snapshotId SnapshotId
+     * @return new instance without the given snapshot
+     */
+    public IndexMetaDataGenerations withRemovedSnapshot(SnapshotId snapshotId) {
+        final Map<SnapshotId, Map<IndexId, String>> updatedIndexMetaLookup = new HashMap<>(lookup);
+        updatedIndexMetaLookup.remove(snapshotId);
+        final Map<String, String> updatedIndexMetaIdentifiers = new HashMap<>(identifiers);
+        updatedIndexMetaIdentifiers.keySet().removeIf(
+            k -> updatedIndexMetaLookup.values().stream().noneMatch(identifiers -> identifiers.containsValue(k)));
+        return new IndexMetaDataGenerations(updatedIndexMetaLookup, updatedIndexMetaIdentifiers);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(identifiers, lookup);
+    }
+
+    @Override
+    public boolean equals(Object that) {
+        if (this == that) {
+            return true;
+        }
+        if (that instanceof IndexMetaDataGenerations == false) {
+            return false;
+        }
+        final IndexMetaDataGenerations other = (IndexMetaDataGenerations) that;
+        return lookup.equals(other.lookup) && identifiers.equals(other.identifiers);
+    }
+
+    @Override
+    public String toString() {
+        return "IndexMetaDataGenerations{lookup:" + lookup + "}{identifier:" + identifiers + "}";
+    }
+
+    /**
+     * Compute identifier for {@link IndexMetaData} from its index uuid as well as its settings-, mapping- and alias-version.
+     * If an index did not see a change in its settings, mappings or aliases between two points in time then the identifier will not change
+     * between them either.
+     *
+     * @param indexMetaData IndexMetaData
+     * @return identifier string
+     */
+    public static String buildUniqueIdentifier(IndexMetaData indexMetaData) {
+        return indexMetaData.getIndexUUID() + "-" + indexMetaData.getSettingsVersion() + "-"
+            + indexMetaData.getMappingVersion() + "-" + indexMetaData.getAliasesVersion();
```

Comment:
> This still leaves the problematic situation where you close indices in a cluster, and restore an older version of those indices from an older snapshot (while there's a newer snapshot) on top of this, which preserves the index uuid, and thereby creates a parallel history of indexMetaData versions.

I wonder if we could just resolve this by not preserving the index uuid on restore? We don't preserve it when renaming on restore if the index exists.
Just using a new uuid seems like it would solve all of these issues?

Reply:
The problem with not preserving the UUID upon restoring in a closed index is that it does not allow you to then incrementally restore from a snapshot (i.e. scenario where a cluster is periodically mirrored onto another cluster by taking regular snapshots and restoring into closed indices on second cluster)

Reply:
Fair point :)

> Tracking this accurately would require a new type of history UUID at the index level (which has come up before in CCR as well, I'm trying to recall the details) that would be reset when you restore an index from a snapshot.

Let's do this then? :) (meaning Add a history UUID that is set on restore to index metadata)

Reply:
Alright, let's see how that would look like (but in a separate PR).